### ISSUE TITLE
Port twisted.internet.ssl to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6142]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6142|
|Type|enhancement|
|Created|2012-10-16 19:24:35Z|
|Branch|https://github.com/twisted/twisted/tree/ssl-py3-6142|

`twisted.internet.ssl` should run on Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__6142 6142
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_ssl_py3_6142 branches/ssl-py3-6142
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1350415475000000 1350415475000000
changetime__1352211370000000 1352211370000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [36308]) Branching to 'ssl-py3-6142'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

This will include `twisted.test.test_ssl`, but not `twisted.internet.test.test_tls`, which depends on SSL endpoints ([#6157](https://github.com/twisted/twisted/issues/6157)), which in turn depends on this ticket. `test_tls` will have to be done separately as a result.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst set owner to @exarkun|
|-|-|

Ready for review; https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ssl-py3-6142 is running.

  1. Re-enabled various places where SSL was disabled on Python 3.
  1. Usual minor fixes.
  1. Removed a little bit of duplication (e.g `certPath`) in the test code.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks.

  1. `twisted/test/test_ssl.py`
    1. in `generateCertificateFiles`, `FilePath.setContent` is better than open/write/close.
    1. likewise for `FilePath.getContent` in `StolenTCPTestCase.createServer`
    1. `ConnectionLostTestCase.test_bothSidesLoseConnection` does a `transport.write("a")` - ie native instead of byte string.  This should cause a test failure...
    1. Do any of these tests run?  They all skip on the Python 3.3 builder and locally.
  1. Some code being changed has comments about ticket [#6141](https://github.com/twisted/twisted/issues/6141) - not [#6142](https://github.com/twisted/twisted/issues/6142)?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set owner to @exarkun|
|-|-|

1.1, 1.2: Fixed.
1.3: Apparently `SSL.Connection.send` will accept unicode. I added a hack to twisted.protocols.tls that will catch this. I also fixed the test.
1.4: They all run... if you have pyOpenSSL installed. The buildslave doesn't, but you can install it locally. I'll email Barry about improving our buildslave setup.

2: Those comments were somewhat inaccurate, the imports should probably be from the public module (i.e. `twisted.internet.ssl`), not `_sslverify`. In any case I missed them when porting `_sslverify` ([#6141](https://github.com/twisted/twisted/issues/6141)), and doing it here is reasonable since this is the public API exposing `_sslverify`.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ssl-py3-6142 is running.
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ssl-py3-6142

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @itamarst|
|-|-|

Okay, thanks.  Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst set status to `closed`|
|-|-|

(In [36350]) Merge ssl-py3-6142.

Author: itamar
Review: exarkun
Fixes: [#6142](https://github.com/twisted/twisted/issues/6142)

Port twisted.internet.ssl to Python 3.
